packages/state-transition/src/cache/epochCache.ts

```diff
@@ -838,3 +901,9 @@ export function createEmptyEpochCacheImmutableData(
     index2pubkey: [],
   };
 }
+
+export function createEmptyCarryoverData(): CarryoverData {
```

Comment:
> Could you explain in detail the rationale of this new `CarryoverData` struct?
> 
> * How does it differentatiate from the rest of caches and sidecars?
> * Why must this be a separate object to existing ones?
>   In case of doubt it's best to not change APIs to minimize the diff and cognitive load to mantainers

With the new unfinalized cache introduced in `EpochCache`, I was thinking if 1) `EpochCache.createFromState()` should allow callers passing in an existed unfinalized cache from other state or should 2) `createFromState()` create an empty unfinalized cache every time. 

I checked every caller of `createFromState()`, and it seems like we are only creating `EpochCache` from a “fresh” state ie. no unfinalized cache. However we cannot guarantee if future development will demand creating `EpochCache` from a `CachedBeaconState`. It wouldn’t make sense to have a new instance of `EpochCache` with empty unfinalized cache in this case. 

I leaned towards the second option. Instead of having the populating logic specifically for unfinalized cache, I think it makes more sense to have a generic container `CarryoverData` that contains any data including `unfinalizedPubkey2index` that need to be populated during `createFromState`. It is almost identical to `EpochCacheImmutableData`, in fact it would be perfectly fine if we put `unfinalizedPubkey2index` in `EpochCacheImmutableData` but this would defeat the purpose as it is unfinalized and mutable hence a separate container. 

I am actually open for 1) because we can always make the additional change later on when we encounter the need for it.

Reply:
Got it. You should do 1 to mantain the same assumptions of the EpochCache design, otherwise it feels like scope creep. I would prefer to discuss changes on the EpochCache usage assumptions in another thread (issue)